extensions/amp-form/0.1/amp-form.js

```diff
@@ -16,14 +16,82 @@
 
 import {isExperimentOn} from '../../../src/experiments';
 import {getService} from '../../../src/service';
+import {assertHttpsUrl} from '../../../src/url';
+import {user} from '../../../src/log';
+import {onDocumentReady} from '../../../src/document-ready';
+import {xhrFor} from '../../../src/xhr';
+import {toArray} from '../../../src/types';
+import {startsWith} from '../../../src/string';
+
 
 /** @type {string} */
 const TAG = 'amp-form';
 
+class AmpForm {
+
+  /**
+   * Adds functionality to the passed form element and listens to submit event.
+   * @param element {!HTMLFormElement}
+   */
+  constructor(element) {
+    /** @const @private {!Window} */
+    this.win_ = element.ownerDocument.defaultView;
+
+    /** @const @private {!Element} */
+    this.form_ = element;
+
+    /** @const @private {!Xhr} */
+    this.xhr_ = xhrFor(this.win_);
+
+    /** @const @private {string} */
+    this.method_ = this.form_.getAttribute('method') || 'GET';
+
+    /** @const @private {string} */
+    this.xhrAction_ = this.form_.getAttribute('data-xhr-action');
+    if (this.xhrAction_) {
+      assertHttpsUrl(this.xhrAction_, this.form_, 'data-xhr-action');
+      user.assert(!startsWith(this.xhrAction_, 'https://cdn.ampproject.org'),
+          'form data-xhr-action should not be on cdn.ampproject.org: %s',
+          this.form_);
+    }
+    this.installSubmitHandler_();
+  }
+
+  /** @private */
+  installSubmitHandler_() {
+    this.form_.addEventListener('submit', e => {
+      if (e.defaultPrevented) {
+        return;
+      }
+      if (this.xhrAction_) {
+        this.xhr_.fetchJson(this.xhrAction_, {
+          body: new FormData(this.form_),
+          method: this.method_,
+        });
+        e.preventDefault();
+      }
+    });
+  }
+}
+
+
+/**
+ * Installs submission handler on all forms in the document.
+ * @param win {!Window}
+ */
+function installSubmissionHandlers(win) {
+  onDocumentReady(win.document, () => {
```

Comment:
I think it's fine as a first pass. But the idea to lazy-instantiate forms based on "submit" event is also interesting as long as it'd not conflict with the global submit listener in runtime. The question, however, is whether we can do the same for input change events since most of the upcoming work will be the area of input validation, messaging, etc.